### ISSUE TITLE
Components are recreated incorrectly for immutable nested arrays

### ISSUE THREAD

**Issue by mickylad** [angular/angular#10182]
**I'm submitting a ...**
```
[x] bug report
[ ] feature request
[ ] support request
```

**Current behavior**
When an item is modified in an immutable array of data, only the component representing that item is recreated - not the components representing the array or any other items.

When an item is modified in an immutable array of arrays of data, the component representing the sub-array that the item is part of and the components representing all of its items are recreated.

**Expected/desired behavior**
When an item is modified in an immutable array of arrays of data, only the component representing that item should be recreated - not the components representing the sub-array or any other items.

**Reproduction of the problem**
[Single list](http://plnkr.co/edit/vnjebqrROzFDsft5lzRW?p=preview)
[List of lists](http://plnkr.co/edit/X566YbJjsIAPVMiGEjgl?p=preview)

Enter a new name for any item using the inputs, then view the logs in the console to see when `ngOnInit` and `ngOnChanges` runs for each component.

**What is the motivation / use case for changing the behavior?**
When representing a matrix of data, performance is poor because more component and DOM changes occur than should be necessary.

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.4

**Notes**
I realise that this may not necessarily be a bug. It's very possible that I simply don't understand how Angular2 is supposed to work. If that's the case then it seems strange that Angular2 would know not to recreate components for immutable data only in certain sitatuations, and there must be some way to represent a matrix to achieve the desired behaviour - but I've yet to find anything that looks like it would help.

I did ask about this [on Gitter](https://gitter.im/angular/angular?at=578f3c412d04e0585195118f), but I've not had any response yet.

**Comment by vicb** [comment]
Please describe what the current behavior is and what you expect  more accurately. Thanks.

**Comment by mickylad** [comment]
**Current behavior**
Supposed I have a service `ListService` which exposes an immutable array of items. In my `AppComponent`, I bind the service's items to a `ListComponent`, which iterates through each item in the array using `ngFor`, binding each to a `ListItemComponent`.

The `ListItemComponent` allows modification of one property of the item bound to it using an `input`. When the input is changed, the event is propagated to the `ListService`, which modifies the item and list (causing a new list to be created with a new item in the index of the modified item).

Angular runs change detection in response to the event, and despite the list reference changing, it seems to know that the `ListComponent` should not be recreated, and simply runs `ngOnChanges` on it. It also knows not to recreate or call `ngOnChanges` on any of the other 

When an item is modified in an immutable array of data, only the component representing that item is recreated - not the components representing the array or any other items.

When an item is modified in an immutable array of arrays of data, the component representing the sub-array that the item is part of and the components representing all of its items are recreated.

**Expected/desired behavior**
When an item is modified in an immutable array of arrays of data, only the component representing that item should be recreated - not the components representing the sub-array or any other items.

**Comment by zoechi** [comment]
@mickylad Did you actually intend to close the issue?

**Comment by mickylad** [comment]
@zoechi, woops, nope! Coming back to this :)

**Comment by mickylad** [comment]
**Current behavior**
Suppose I have a `ListService` that exposes an immutable two-dimensional array. In my `AppComponent`, I bind that 2D array to a `ListListComponent`, which uses two nested `ngFor`s to iterate through the items in each inner array, and bind each one to a `ListItemComponent`.

Inside the `ListService`, I then modify one item, which means the references to that item, the inner array that contains it, and the 2D array all change. Angular calls `ngOnChanges` on the `ListListComponent`, then creates new `ListItemComponent`s for *all* of the items in the inner array, despite only one having changed.

The creation of new `ListItemComponent`s for the unmodified items seems unnecessary, and is expensive for large data sets.

**Expected/desired behavior**
I don't have a full understanding of how changes are handled with regards to the component lifecycle, but the desired behaviour would be one of the following:

1. When the item is modified, `ngOnChanges` is called on the `ListListComponent`, and a new `ListItemComponent` is created for *only* the item that changed.
2. When the item is modified, `ngOnChanges` is called on the `ListListComponent`, the `ListItemComponent` that changed, and nothing else. 

In either way, the key point is that new `ListItemComponent`s should not be created for the unmodified items in the modified inner array.

I've created another Plunker, with the above case implemented: http://plnkr.co/edit/fz4sFBNmPQ4PuogHYuNs?p=preview.

Each item displays the last call to either `ngInit` or `ngOnChanges` with a timestamp, and the first item in the first inner array is modified every 500ms.

**Comment by mickylad** [comment]
Hopefully that is a bit clearer!

**Comment by awerlang** [comment]
Interesting PoC.

I modified your plunker slightly to show it working with a mutating list, but immutable objects and OnPush on ListItemComponent. Seems to be the best combination currently.

**Comment by mickylad** [comment]
Thanks @awerlang !

I did investigate the use of `OnPush`. I understand it affects change detection, so that component branches are ignored unless the `@Input()`s change, and it is something I would like to make use of in general. However, it didn't seem to actually help in my situation.

It is still left in the two Plunkers from my original comment, but removing it doesn't affect their behaviour. I guess because it's not a problem with incorrectly detecting changes, but with tracking the nested components when changes _are_ detected. I couldn't see how it affected your Plunker either - do you see different behaviour if you remove it from `ListItemComponent`?

Using mutable structures is an option, although one I was trying to avoid. It restricts the use of `OnPush`, which we can see in your Plunker - adding it back into `ListListComponent` will stop changes being detected in the nested components.

This restriction would be applied to _every_ component that uses the same piece of data. If we have multiple sets of data that are affected by this, we're going to end with a lot of mutable data floating round, and we'll have to use `OnPush` very sporadically.

**Comment by awerlang** [comment]
> I did investigate the use of OnPush. I understand it affects change detection, so that component branches are ignored unless the @Input()s change, and it is something I would like to make use of in general. However, it didn't seem to actually help in my situation.

> It is still left in the two Plunkers from my original comment, but removing it doesn't affect their behaviour. I guess because it's not a problem with incorrectly detecting changes, but with tracking the nested components when changes are detected. I couldn't see how it affected your Plunker either - do you see different behaviour if you remove it from ListItemComponent?

I agree with you, when angular decides it needs to update a component, its children seems to be recreated everytime. Removing `OnPush` doesn't cause a visual difference, just the contents of ListItemComponent would have to be _checked_ everytime. `ListItemComponent` doesn't have any input, not sure if that's an edge case.

It is stated that the usage of immutable structures would make things faster, but it really avoids change detection probes at the expense of more changes to the DOM as the effects of a change detection that determined that data has changed.

> Using mutable structures is an option, although one I was trying to avoid. It restricts the use of OnPush, which we can see in your Plunker - adding it back into ListListComponent will stop changes being detected in the nested components.

> This restriction would be applied to every component that uses the same piece of data. If we have multiple sets of data that are affected by this, we're going to end with a lot of mutable data floating round, and we'll have to use OnPush very sporadically.

The restriction is: every child in an OnPush component must be OnPush too, these components doesn't have a choice. Being unable to using OnPush is because is just not compatible with mutable structures.

**Comment by mickylad** [comment]
> The restriction is: every child in an OnPush component must be OnPush too, these components doesn't have a choice. Being unable to using OnPush is because is just not compatible with mutable structures.

Sure, but that's my problem - I don't want to introduce mutable structures to my state, nor end up using `OnPush` only in selected components. That's going to be a nightmare to keep track of.

> It is stated that the usage of immutable structures would make things faster, but it really avoids change detection probes at the expense of more changes to the DOM as the effects of a change detection that determined that data has changed.

Agreed, and I'd go as far to say that because of the poor performance when handling those changes, Angular2 is not compatible with immutable data structures in complex applications right now. But I don't think it has to be that way. It should be possible to keep track of references of child components in a smarter way.

**Comment by tomwanzek** [comment]
I just experienced the essentially same problem trying to debug it. The setup is based on ngrx/observables.

**TL;DR: The solution is using `trackBy` with a tracking function in the `ngFor`. I.e. a unique identifier within the immutable object over which ngFor iterates.** Thanks to the @vicb and @sky-coding for the hint in #11727 🎉 

The main reason, why I am going to add a little more here, is the following.

In my case the immutable array for a store slice was fed as an observable  using `ngFor` and `async` pipe into a custom component input. The custom component in turn has child components which take properties of the input immutable as inputs (again each  immutable with `onPush`). They have some reactive forms and emit change events back up the component hierarchy, which are dispatched back to store for reduction... full circle.

The **unusual and hard to debug** implication of the issue, was that both Chrome and Edge crashed the app somewhere in/after the child components `ngOnChanges` event handler. I.e. no stack traces, console.logs or the like before angular-cli hot reloaded. Only Firefox showed that every hit to `ngOnChange` was essentially a brand new render (`previousValue` uninitialized and `isFirstChange()` always `true`).

The above behavior occurred regardless of whether the change detection strategy was `onPush` or `default` (used the later when I tried to debug). Not sure, what caused the crashes? I suspect, it might have been something about the interaction with the reactive formGroups/formControls in the child components.

**But `trackBy` fixed it across all browsers and only the nested "new" immutable parts of the overall data structure get refreshed with `onPush`**

**As for the already amazing docs on angular.io:**
* it might be worthwhile to call out _immutables_ explicitly in the ngFor documentation (the content about DOM teardown is there already, it might just be the extra prominence of  the use case),
* additionally callout the implications of using ngFor in conjunction with immutables in the Lifecycle Hooks section of the Advanced topics ('cautionary tale' warning) under `OnChange`
* in the  API docs under change  detection strategy

It's the last two bullets, where I checked first to figure out what was going on, as the crashes occurred when hitting the child components which are not in an ngFor iteration.

With that being said, I am glad I found this issue and the solution under #11727 👍

**Comment by vicb** [comment]
As @tomwanzek said above, the solution is to use `trackBy`
